### DISCUSSION TITLE
Is it possible to have multiple schedulers and still benefit from dependency injection?

### DISCUSSION THREAD

**Discussion by tarockx** [quartznet/quartznet#1814] [Q&A]
Hi everyone

I'm migrating a legacy application that uses Quartz to an ASP.NET Core web app, and in this application I need to have multiple `IScheduler` instances.

In the legacy app I achieved this easily by doing this for every scheduler:

```
StdSchedulerFactory factory = new StdSchedulerFactory(props);
scheduler = await factory.GetScheduler().Result;
```

However, in .NET Core I would like to take advantage of Dependency Injection integration so I can inject services into the jobs.
Is this possible while still having the ability to create multiple schedulers? I saw that I can then get `ISchedulerFactory` from the service container, however the factory doesn't seem to have a method to create new schedulers, it can just retrieve the one you initially configure when calling `.AddQuartz()` .

Is this possible? How can I do it?
Thanks!

**Comment by lahma** [comment]
Currently only one is supported. But you could probably create a workaround by manually registering `ISchedulerFactory` instances and utilizing `SchedulerBuilder` to create them. `SchedulerBuilder` has pretty much the same API available. You also need to change the job factory accordingly the MS DI one.

**Reply by oab-ms** [reply]
Unfortunately It won't be enough if you also need to specify a component such as a IJobFactory `.UseJobFactory<T>()` where T does not have a default constructor (because it also depends on the unity container).

**Comment by tarockx** [comment]
Thanks for your answer, however I'm afraid I'm not entirely sure how to put it into practice 🤔

I tried injecting `MicrosoftDependencyInjectionJobFactory` into DI and I see it works, I can get the instance from DI without problems.
However, when creating the builder I don't understand how to pass the instance of the job factory to it, since I can only specify the type:
```
var builder = SchedulerBuilder.Create();
var factory = builder.WithId(myId)
    .UseJobFactory<MicrosoftDependencyInjectionJobFactory>() //cannot pass instance here, just specify type
    .Build();
```

and obviously this gives me this error:
`Unable to Instantiate JobFactory: Cannot instantiate type which has no empty constructor (Parameter 'MicrosoftDependencyInjectionJobFactory')`
since SchedulerBuilder is not part in the DI chain.

So I also tried directly also injecting `SchedulerBuilder` into DI, but that doesn't work either because it doesn't have a public constructor so DI can't instantiate it, I get an error right away.

Can you kindly point me in the right direction?
Thank you in advance

**Reply by oab-ms** [reply]
Having the same exact issue here with the JobFactory. Is there a way of passing an instantiated object instead?
Is there a workaround?

**Comment by tarockx** [comment]
Is there anyone that can give me some further pointers on this? 😥

**Reply by xantari** [reply]
Curious, why would you need more than one scheduler?

**Reply by oab-ms** [reply]
In our case we had a regional regulatory requirement that we were moving towards and that impacted our scheduled jobs. Originally I was working to have a flighted old/new as separate implementations with their own schedulers. This would have made clean up easier after the new one was found to work reliably in production. Injecting our custom IJobFactory implementation would not work with creating a new scheduler, at least the way I did it.

As a workaround, I have them all in one implementation now and I essentially put the scheduler on standby, delete all the jobs in the default scheduler before adding the new ones and then on a separate flag we enable to other after some time has passed so the default scheduler can start up again. So far so good, but it hasn't been tested well enough to say if it will work.

Given that existing jobs will still be running (I am aware there is a config to cause active jobs to get deleted but it might lead to side effects) we have a separate flight for enabling non-regionalized vs regionalized tasks.

**Reply by lahma** [reply]
I think it might make sense to have something like `.AddQuartz(x => x.AddScheduler<ICustomFactory>("name", "id", opt => {}))` which would allow you to configure named and interface-bound scheduler on top of default scheduler.

**Reply by tarockx** [reply]
> Curious, why would you need more than one scheduler?

Because I'm developing a multi-tenant application where each tenant has a dedicated scheduler that runs their tasks independently. Having one scheduler per tenant allows me to easily stop/pause/restart/reconfigure one of them without interfering with the others, and also allows me to have different settings for every tenant (example: a tenant might be allowed to have multiple execution threads for their jobs while another might only get one). You could theoretically implement all this with one scheduler and a careful use of observers, semaphores queues, etc. but it would be much more complex than simply having separate schedulers.
